src/Storage/Notmuch.hs

```diff
@@ -51,28 +52,38 @@ instance ManageTags NotmuchThread where
   tags = thTags
   writeTags = setNotmuchThreadTags
 
+-- | A helper function for opening, performing work,
+-- and closing the database
+withDatabase
+  :: forall a m c. (Notmuch.Mode a, MonadError Error m, MonadIO m)
+  => FilePath
+  -> (Notmuch.Database a -> ExceptT Error IO c)
+  -> m c
+withDatabase dbpath = bracketT (Notmuch.databaseOpen dbpath) Notmuch.databaseDestroy
 
 -- | creates a vector of parsed mails from a not much search
 -- Note, that at this point in time only free form searches are supported. Also,
 -- we filter out the tag which we use to mark mails as new mails
 getMessages
-  :: (MonadError Error m, MonadIO m)
+  :: forall m. (MonadError Error m, MonadIO m)
   => T.Text
   -> NotmuchSettings FilePath
   -> m (Vec.Vector NotmuchMail)
 getMessages s settings =
-  bracketT (Notmuch.databaseOpenReadOnly (view nmDatabase settings)) Notmuch.databaseDestroy go
+  let withDb = withDatabase (view nmDatabase settings) :: (Notmuch.Database Notmuch.RO -> ExceptT Error IO c) -> m c
+  in withDb go
```

Comment:
Instead of using `ScopedTypeVariables` and these type signatures which are... a bit heavy, let's create a variant
of `withDatabase` specialised to `RO` database.  So:
```haskell
withDatabase
  :: (Notmuch.AsNotmuchError e, Notmuch.Mode a, MonadError e m, MonadIO m)
  => FilePath
  -> (Notmuch.Database a -> ExceptT e IO c)
  -> m c
withDatabase dbpath = bracketT (Notmuch.databaseOpen dbpath) Notmuch.databaseDestroy

withDatabaseReadOnly
  :: (Notmuch.AsNotmuchError e, MonadError e m, MonadIO m)
  => FilePath
  -> (Notmuch.Database Notmuch.RO -> ExceptT e IO c)
  -> m c
withDatabaseReadOnly = withDatabase
```

Then we can just use `databaseOpenReadOnly` where appropriate, and avoid `ScopedTypeVariables`.

Reply:
Thanks for pointing this out!